File: Kafka-Utils-Java/src/main/java/com/yhyr/comsumer/ConsumerSpecialOffset.java
package com.yhyr.comsumer;

import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;

import java.util.Collections;
import java.util.Properties;

/**
 * Class Subject: 从指定Offset处开始消费
 * <p>
 * Tips: Consumer默认会自动帮你提交Offset信息(enable.auto.commit=true); 如果只想基于特定Offset做数据嗅探,
 * 而不改变broker端原始的offset信息, 可以指定enable.auto.commit=false;
 *
 * @author yhyr
 * @since 2018/12/30 18:24
 */
public class ConsumerSpecialOffset {
    public static void main(String[] args) {
        String brokers = "localhost:9092";
        String topic = "topic_demo";
        String group = "group_test";

        int customPartitionOffset = 500;

        Properties props = new Properties();
        props.put("bootstrap.servers", brokers);
        props.put("group.id", group);
        props.put("enable.auto.commit", "false");
        props.put("auto.offset.reset", "latest");
        props.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        props.put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props);

        /**
         * KafkaConsumer.seek 是基于客户端的行为;
         *
         * 因此在订阅Topic后, 必须调用一次poll方法, 用来获取Client被分配的Partition, 方可通过seek重新指定消费位置
         */
        consumer.subscribe(Collections.singletonList(topic));
        consumer.poll(0);
        consumer.assignment().forEach(topicPartition -> consumer.seek(topicPartition, customPartitionOffset));

        while (true) {
            ConsumerRecords<String, String> records = consumer.poll(100);
            for (ConsumerRecord<String, String> record : records) {
                System.out.println(record.toString());
            }
        }
    }
}

File: Kafka-Utils-Python/consumer/consumer_special_offset.py
# -*- coding: utf-8 -*-
"""
Subject: 从指定Offset处开始消费消息
Tips:
    Kafka默认会自动提交Offset信息(enable_auto_commit=True),
    如果只是想基于特定Offset做数据嗅探, 而不改变broker端原始的offset信息, 可以指定enable_auto_commit=False

    消息的消费是基于具体的Partition而言, 所以指定Offset的值实际是作用于具体的一个或多个Partition, 而非是Topic

@Author YH YR
@Time 2018/12/31 14:09
"""

from kafka import KafkaConsumer


class ConsumerSpecialOffset:
    def __init__(self, broker_list, group_name, topic, enable_auto_commit=True, auto_offset_reset='latest'):
        self.broker_list = broker_list
        self.topic = topic
        self.group_name = group_name
        self.enable_auto_commit = enable_auto_commit
        self.auto_offset_reset = auto_offset_reset

    def consumer_from_offset(self, process_msg, offset):
        consumer = KafkaConsumer(group_id=self.group_name, bootstrap_servers=self.broker_list,
                                 enable_auto_commit=self.enable_auto_commit, auto_offset_reset=self.auto_offset_reset)
        consumer.subscribe(self.topic)
        consumer.poll(0)
        for topic_partition in consumer.assignment():
            consumer.seek(topic_partition, offset)
        while True:
            consumer_records = consumer.poll(100)
            for partition_info, records in consumer_records.items():
                for record in records:
                    process_msg(record)


def print_msg(msg_dic):
    print(msg_dic)


if __name__ == '__main__':
    broker_list = 'localhost:9092'
    group_name = 'group_test'
    topic = ['topic_demo']

    action = ConsumerSpecialOffset(broker_list, group_name, topic, enable_auto_commit=True)
    action.consumer_from_offset(print_msg, 1000)
